agent/cache-types/catalog_list_services.go

```diff
@@ -38,15 +38,23 @@ func (c *CatalogListServices) Fetch(opts cache.FetchOptions, req cache.Request)
 	// going to be served from cache and end up arbitrarily stale anyway. This
 	// allows cached service-discover to automatically read scale across all
 	// servers too.
-	reqReal.AllowStale = true
+	reqReal.QueryOptions.AllowStale = true
+
+	if opts.LastResult != nil {
+		reqReal.QueryOptions.AllowNotModifiedResponse = true
+	}
 
-	// Fetch
 	var reply structs.IndexedServices
 	if err := c.RPC.RPC("Catalog.ListServices", reqReal, &reply); err != nil {
 		return result, err
 	}
 
-	result.Value = &reply
 	result.Index = reply.QueryMeta.Index
+
+	if reply.QueryMeta.NotModified {
+		result.Value = opts.LastResult.Value
+		return result, nil
+	}
+	result.Value = &reply
```

Comment:
This is a bit different from #7974, which checked this value in the generic part of the cache. 

Handling `NotModified` here removes the need to add another field to `FetchResult`, at the cost of having to implement the logic in every application `cache-type`. Since not every endpoint may support this field it maybe doesn't hurt to be explicit about which cache entries are using it.

Reply:
Ok, I was more about have the same mechanism for all (hence the additional field in FetchResult) but you are right, that's around the same amount of code